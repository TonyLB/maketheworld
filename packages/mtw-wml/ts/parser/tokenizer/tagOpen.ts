import { Tokenizer, TokenTagOpenBegin, TokenTagOpenEnd } from "./baseClasses"

export const tagOpenBeginTokenizer: Tokenizer<TokenTagOpenBegin> = (sourceStream) => {
    if (sourceStream.lookAhead('<')) {
        const startIdx = sourceStream.position
        let returnValue = sourceStream.consume(1)
        let tag = ''
        while(sourceStream.lookAhead(1).match(/[A-Za-z]/)) {
            tag = tag + sourceStream.consume(1)
        }
        returnValue = returnValue + tag
        if (returnValue) {
            return {
                type: 'TagOpenBegin',
                source: returnValue,
                tag: tag,
                startIdx,
                endIdx: sourceStream.position - 1
            }
        }
    }
    return undefined
}

export const tagOpenEndTokenizer: Tokenizer<TokenTagOpenEnd> = (sourceStream) => {
    const startIdx = sourceStream.position
    if (sourceStream.lookAhead('>')) {
        return {
            type: 'TagOpenEnd',
            source: sourceStream.consume(1),
            startIdx,
            endIdx: sourceStream.position - 1,
            selfClosing: false
        }
    }
    if (sourceStream.lookAhead('/>')) {
        return {
            type: 'TagOpenEnd',
            source: sourceStream.consume(2),
            startIdx,
            endIdx: sourceStream.position - 1,
            selfClosing: true
        }
    }
    return undefined
}
