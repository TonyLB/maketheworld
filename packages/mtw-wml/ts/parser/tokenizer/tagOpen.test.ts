import SourceStream from './sourceStream'
import { tagOpenBeginTokenizer, tagOpenEndTokenizer } from './tagOpen'

describe('tagOpenBeginTokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return undefined when no match', () => {
        const testStream = new SourceStream('Testing')
        expect(tagOpenBeginTokenizer(testStream)).toBe(undefined)
    })
    it('should tokenize tag open', () => {
        const testStream = new SourceStream('<Room key=(ABC)>')
        expect(tagOpenBeginTokenizer(testStream)).toMatchSnapshot()
    })
})

describe('tagOpenEndTokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return undefined when no match', () => {
        const testStream = new SourceStream('Testing')
        expect(tagOpenEndTokenizer(testStream)).toBe(undefined)
    })
    it('should tokenize non-self-closing tag close', () => {
        const testStream = new SourceStream('></Room>')
        expect(tagOpenEndTokenizer(testStream)).toMatchSnapshot()
    })
    it('should tokenize self-closing tag close', () => {
        const testStream = new SourceStream('/></Room>')
        expect(tagOpenEndTokenizer(testStream)).toMatchSnapshot()
    })
})