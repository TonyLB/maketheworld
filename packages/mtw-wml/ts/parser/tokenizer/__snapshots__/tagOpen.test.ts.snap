// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tagOpenBeginTokenizer should tokenize tag open 1`] = `
Object {
  "endIdx": 4,
  "source": "<Room",
  "startIdx": 0,
  "tag": "Room",
  "type": "TagOpenBegin",
}
`;

exports[`tagOpenEndTokenizer should tokenize non-self-closing tag close 1`] = `
Object {
  "endIdx": 0,
  "selfClosing": false,
  "source": ">",
  "startIdx": 0,
  "type": "TagOpenEnd",
}
`;

exports[`tagOpenEndTokenizer should tokenize self-closing tag close 1`] = `
Object {
  "endIdx": 1,
  "selfClosing": true,
  "source": "/>",
  "startIdx": 0,
  "type": "TagOpenEnd",
}
`;
